### PR TITLE
add Sons of the Forrest

### PR DESCRIPTION
**Pull request by gOOvER**
# Description

add Egg for Sons of the Forrest

fixes: #2318 

## Checklist for all submissions

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/parkervcp/eggs/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?
* [x] Have you tested and reviewed your changes with confidence that everything works?
* [x] Did you branch your changes and PR from that branch and not from your master branch?
  * If not, why?:

## New egg Submissions

1. [x] Does your submission pass tests (server is connectable)?
3. [x] Have you added the egg to the main README.md and any other README files in subdirectories of the egg (e.g /game_eggs) according to the alphabetical order?
4. [x] Have you added a unique README.md for the egg you are adding according to the alphabetical order?
5. [x] You verify that the start command applied does not use a shell script
    * [ ] If some script is needed then it is part of a current yolk or a PR to add one
6. [x] The egg was exported from the panel

### CODE REVIEW

**Comment by gOOvER** [comment]
ahhh wait; wrong images :D fix is asap :D

**Comment by gOOvER** [comment]
I will update the urls for default configs, when the egg is ready to merge.

**Comment by QuintenQVD0** [comment]
If someone can confirm it works then this can be merged.

**Comment by vp-en** [comment]
I have tested it :). It works great out of the box.

**Comment by gOOvER** [comment]
@vp-en thats a ptero bug

**Comment by vp-en** [comment]
> @vp-en thats a ptero bug. you can only use ^^C atm

Yeah alright, fair enough :). Would be nice to have some kind of supervisor built into Pterodactyl, like the s6-supervisor. Anyways, not related to this PR 😄 

This egg works great. 

Not sure if it's possible, but it would be a cool addition, if some of the "[Game Options](https://steamcommunity.com/sharedfiles/filedetails/?id=2992700419#6333179)" could be toggleable in the startup variables. Like the `Gameplay.TreeRegrowth` and `Structure.Damage`

**Comment by gOOvER** [comment]
> > @vp-en thats a ptero bug. you can only use ^^C atm
> 
> Yeah alright, fair enough :). Would be nice to have some kind of supervisor built into Pterodactyl, like the s6-supervisor. Anyways, not related to this PR 😄
> 
> This egg works great.
> 
> Not sure if it's possible, but it would be a cool addition, if some of the "[Game Options](https://steamcommunity.com/sharedfiles/filedetails/?id=2992700419#6333179)" could be toggleable in the startup variables. Like the `Gameplay.TreeRegrowth` and `Structure.Damage`

i updated my comment with some more infos :)

File: game_eggs/README.md [hunk 1]
```diff
@@ -171,6 +171,7 @@
   * [Exiled](steamcmd_servers/scpsl/exiled)
   * [Multiadmin](steamcmd_servers/scpsl/multiadmin)
 * [Soldat](steamcmd_servers/soldat)
+* [Sons of the Forrest](steamcmd_servers/sonsoftheforrest)
```
Comment (rathmerdominik) [inline game_eggs/README.md:174]: Little grammar mistake here
```suggestion
* [Sons of the Forest](steamcmd_servers/sonsoftheforest)
```

File: README.md [hunk 2]
```diff
@@ -271,6 +271,7 @@ If you are reading this it looks like you are looking to add an egg to your serv
   * [exiled](game_eggs/steamcmd_servers/scpsl/exiled)
   * [multiadmin](game_eggs/steamcmd_servers/scpsl/multiadmin)
 * [Soldat](game_eggs/steamcmd_servers/soldat)
+* [Sons of the Forrest](game_eggs/steamcmd_servers/sonsoftheforrest)
```
Comment (rathmerdominik) [inline README.md:274]:
```suggestion
* [Sons of the Forest](game_eggs/steamcmd_servers/sonsoftheforest)
```

File: game_eggs/steamcmd_servers/README.md [hunk 3]
```diff
@@ -189,6 +189,10 @@ This is a collection of servers that use SteamCMD to install.
 
 [Soldat](soldat)
 
+## Sons of the Forrest
+
+[Sons of the Forrest](sonsoftheforrest)
```
Comment (rathmerdominik) [inline game_eggs/steamcmd_servers/README.md:194]:
```suggestion
[Sons of the Forest](sonsoftheforest)
```

File: game_eggs/steamcmd_servers/sonsoftheforrest/README.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
+# Sons of the Forrest
+
+### Game Description
+
+Sons of the Forest is a horror survival game and sequel to The Forest by Endnight Games, Ltd.. Sent to find a missing billionaire on a remote island, you find yourself in a cannibal-infested hellscape. Craft, build, and struggle to survive, alone or with friends.
+
+### Usefull links
```
Comment (rathmerdominik) [inline game_eggs/steamcmd_servers/sonsoftheforrest/README.md:7]:
```suggestion
### Useful links
```

File: game_eggs/steamcmd_servers/sonsoftheforrest/README.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+### Author & Contributers
+| Name        | Github Profile  | Buy me a Coffee |
+| ------------- |-------------|-------------|
+|   gOOvER   | https://github.com/gOOvER | [![ko-fi](https://ko-fi.com/img/githubbutton_sm.svg)](https://ko-fi.com/B0B351D0Q) |
+
+
+### Server Ports
+
+Sons of the Forrest requires up to 3 ports. You can choose every port you want.
```
Comment (rathmerdominik) [inline game_eggs/steamcmd_servers/sonsoftheforrest/README.md:19]:
```suggestion
Sons of the Forest requires up to 3 ports. You can choose every port you want.
```

File: game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+            "field_type": "text"
+        },
+        {
+            "name": "Password",
+            "description": "",
+            "env_variable": "SRV_PW",
+            "default_value": "changeme",
+            "user_viewable": true,
+            "user_editable": true,
+            "rules": "nullable|string",
```
Comment (rathmerdominik) [inline game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json:80]: Maximum of 40 as stated in the documentation

File: game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+            "field_type": "text"
+        },
+        {
+            "name": "Max Players",
+            "description": "The maximum number of players allowed simultaneously on the server.",
+            "env_variable": "MAX_PLAYERS",
+            "default_value": "4",
+            "user_viewable": true,
+            "user_editable": true,
+            "rules": "required|integer|min:1",
```
Comment (rathmerdominik) [inline game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json:70]: Maximum of 8 Players allowed

File: game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+            "field_type": "text"
+        },
+        {
+            "name": "Server Name",
+            "description": "Name of the server visible in the server list, and in the Steam contacts.",
+            "env_variable": "SRV_NAME",
+            "default_value": "Pterodactyl hosted Server",
+            "user_viewable": true,
+            "user_editable": true,
+            "rules": "required|string|max:60",
```
Comment (rathmerdominik) [inline game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json:60]: Not sure if there is a real limit required. The documentation only says "Any string".

Is there a specific Limit given?

**Review by rathmerdominik — CHANGES_REQUESTED** [review]
A few grammar issues :smile: 
But else looks good.

**Comment by gOOvER** [comment]
> A few grammar issues 😄 But else looks good.

stop with that smilie ;) not for everyone is english native ;)

File: game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json (hunk 6 above)
Comment (gOOvER) [reply game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json:80]: pls provide a link ;)

File: game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json (hunk 7 above)
Comment (gOOvER) [reply game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json:70]: ple provide a link ;)

File: game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json (hunk 8 above)
Comment (gOOvER) [reply game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json:60]: > Not sure if there is a real limit required. The documentation only says "Any string".
> 
> Is there a specific Limit given?

shgow me, that more is allowed ;)

**Comment by rathmerdominik** [comment]
> > A few grammar issues smile But else looks good.
> 
> stop with that smilie ;) not for everyone is english native ;) I am very happy about reports of incorrect spelling, but there is no need to make fun of it.

Uhh I did not intend to make fun of you :( i am german myself so i am also not native. It was more a "Nice job!" Smile :smile:

File: game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json (hunk 7 above)
Comment (rathmerdominik) [reply game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json]: https://steamcommunity.com/sharedfiles/filedetails/?id=2992700419&snr=1_5_9_ here it is :)

File: game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json (hunk 8 above)
Comment (rathmerdominik) [reply game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json]: https://steamcommunity.com/sharedfiles/filedetails/?id=2992700419&snr=1_5_9_ Here ^^
The section says "any string"

File: game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json (hunk 6 above)
Comment (rathmerdominik) [reply game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json]: https://steamcommunity.com/sharedfiles/filedetails/?id=2992700419&snr=1_5_9_ Here you go!

File: game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json (hunk 8 above)
Comment (rathmerdominik) [reply game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json]: image

File: game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json (hunk 7 above)
Comment (rathmerdominik) [reply game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json]: image

File: game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json (hunk 6 above)
Comment (rathmerdominik) [reply game_eggs/steamcmd_servers/sonsoftheforrest/egg-sons-of-the-forest.json]: image

**Comment by rathmerdominik** [comment]
image
I somtimes get this error when starting the server

This seems to be rather flaky though as it sometimes happens and sometimes it does not

I can reproduce this locally on linux with wine

Some executions run the Server without a fault and others have this exact same error

I am absolutely not sure how one would debug this

**Comment by gOOvER** [comment]
> image
> 
> I somtimes get this error when starting the server
> 
> This seems to be rather flaky though as it sometimes happens and sometimes it does not
> 
> I can reproduce this locally on linux with wine
> 
> Some executions run the Server without a fault and others have this exact same error
> 
> I am absolutely not sure how one would debug this

its not an egg issue; more a firewall issue. Open all ports. I like the Servercheck :)

**Comment by rathmerdominik** [comment]
> > image
> > I somtimes get this error when starting the server
> > This seems to be rather flaky though as it sometimes happens and sometimes it does not
> > I can reproduce this locally on linux with wine
> > Some executions run the Server without a fault and others have this exact same error
> > I am absolutely not sure how one would debug this
> 
> its not an egg issue; more a firewall issue. Open all ports. I like the Servercheck :)


It does not even get to the point where it can check the open ports. This error happens before that and has something to do with the SteamUser becoming unset for whatever reason.

The moments where it gets to the open port check it works flawlessly and the server starts and is reachable.

You are correct though that this is not an issue with the egg but more an issue with the Dedicated Server or the implementation of the ISteamClient.

The ports or firewall are certainly not the problem as it does not even get to the check, as i have already mentioned

**Comment by gOOvER** [comment]
> 



> > > A few grammar issues smile But else looks good.
> > 
> > 
> > stop with that smilie ;) not for everyone is english native ;) I am very happy about reports of incorrect spelling, but there is no need to make fun of it.
> 
> Uhh I did not intend to make fun of you :( i am german myself so i am also not native. It was more a "Nice job!" Smile 😄


> > > image
> > > I somtimes get this error when starting the server
> > > This seems to be rather flaky though as it sometimes happens and sometimes it does not
> > > I can reproduce this locally on linux with wine
> > > Some executions run the Server without a fault and others have this exact same error
> > > I am absolutely not sure how one would debug this
> > 
> > 
> > its not an egg issue; more a firewall issue. Open all ports. I like the Servercheck :)
> 
> It does not even get to the point where it can check the open ports. This error happens before that and has something to do with the SteamUser becoming unset for whatever reason.
> 
> The moments where it gets to the open port check it works flawlessly and the server starts and is reachable.
> 
> You are correct though that this is not an issue with the egg but more an issue with the Dedicated Server or the implementation of the ISteamClient.
> 
> The ports or firewall are certainly not the problem as it does not even get to the check, as i have already mentioned

no, it works now for 6 Hosters with, idk customers; so fix your firewall.

**Comment by rathmerdominik** [comment]
> > 
> 
> > > > A few grammar issues smile But else looks good.
> > > 
> > > 
> > > stop with that smilie ;) not for everyone is english native ;) I am very happy about reports of incorrect spelling, but there is no need to make fun of it.
> > 
> > 
> > Uhh I did not intend to make fun of you :( i am german myself so i am also not native. It was more a "Nice job!" Smile smile
> 
> > > > image
> > > > I somtimes get this error when starting the server
> > > > This seems to be rather flaky though as it sometimes happens and sometimes it does not
> > > > I can reproduce this locally on linux with wine
> > > > Some executions run the Server without a fault and others have this exact same error
> > > > I am absolutely not sure how one would debug this
> > > 
> > > 
> > > its not an egg issue; more a firewall issue. Open all ports. I like the Servercheck :)
> > 
> > 
> > It does not even get to the point where it can check the open ports. This error happens before that and has something to do with the SteamUser becoming unset for whatever reason.
> > The moments where it gets to the open port check it works flawlessly and the server starts and is reachable.
> > You are correct though that this is not an issue with the egg but more an issue with the Dedicated Server or the implementation of the ISteamClient.
> > The ports or firewall are certainly not the problem as it does not even get to the check, as i have already mentioned
> 
> no, it works now for 6 Hosters with, idk customers; so fix your firewall.

I am not really sure if you do not want to read what i write or just type without reading.
It fails **before** even getting to anything net related. 
There is a **runtime error** that crashes the Server and nothing Net Related

Try to restart your server multiple times. Sometimes it will run and sometimes it will not. It is **flaky**.

Flaky means it will work at some runs and at some it will not.

And at the runs where this **runtime error** does not occur the server **works** and is also able to be connected to

image

Here a **working** run

My server is there and you are able to connect to it

**Comment by gOOvER** [comment]
> 

i read it; my server runs now for 14h; since release and 2 hosters with 200+, customers have no problems; The servercheck tell you whats wrong; egg works fine

**Comment by gOOvER** [comment]
> > > 
> > 
> > 
> > > > > A few grammar issues smile But else looks good.
> > > > 
> > > > 
> > > > stop with that smilie ;) not for everyone is english native ;) I am very happy about reports of incorrect spelling, but there is no need to make fun of it.
> > > 
> > > 
> > > Uhh I did not intend to make fun of you :( i am german myself so i am also not native. It was more a "Nice job!" Smile smile
> > 
> > 
> > > > > image
> > > > > I somtimes get this error when starting the server
> > > > > This seems to be rather flaky though as it sometimes happens and sometimes it does not
> > > > > I can reproduce this locally on linux with wine
> > > > > Some executions run the Server without a fault and others have this exact same error
> > > > > I am absolutely not sure how one would debug this
> > > > 
> > > > 
> > > > its not an egg issue; more a firewall issue. Open all ports. I like the Servercheck :)
> > > 
> > > 
> > > It does not even get to the point where it can check the open ports. This error happens before that and has something to do with the SteamUser becoming unset for whatever reason.
> > > The moments where it gets to the open port check it works flawlessly and the server starts and is reachable.
> > > You are correct though that this is not an issue with the egg but more an issue with the Dedicated Server or the implementation of the ISteamClient.
> > > The ports or firewall are certainly not the problem as it does not even get to the check, as i have already mentioned
> > 
> > 
> > no, it works now for 6 Hosters with, idk customers; so fix your firewall.
> 
> I am not really sure if you do not want to read what i write or just type without reading. It fails **before** even getting to anything net related. There is a **runtime error** that crashes the Server and nothing Net Related
> 
> Try to restart your server multiple times. Sometimes it will run and sometimes it will not. It is **flaky**.
> 
> Flaky means it will work at some runs and at some it will not.
> 
> And at the runs where this **runtime error** does not occur the server **works** and is also able to be connected to
> 
> image
> 
> Here a **working** run
> 
> My server is there and you are able to connect to it

then post the RUNTIME Error

**Comment by rathmerdominik** [comment]
> > 
> 
> i read it; my server runs now for 14h; since release and 2 hosters with 200+, customers have no problems; The servercheck tell you whats wrong; egg works fine

image

As you can read in the error message that i have sent you there is no information at all. 
It is a simple crash with an exception message that clearly states `Assertion Failed: Attempt to call interface with invalid hSteamUser`

That is the error. That is what makes the game crash before even getting to the port check

Also a **running** server works completely fine.
I said you need to **restart** your server multiple times
And you will see that at some starts it will not boot but throw this error

thats what i meant with flaky

**Comment by gOOvER** [comment]
> > > 
> > 
> > 
> > i read it; my server runs now for 14h; since release and 2 hosters with 200+, customers have no problems; The servercheck tell you whats wrong; egg works fine
> 
> image
> 
> As you can read in the error message that i have sent you there is no information at all. It is a simple crash with an exception message that clearly states `Assertion Failed: Attempt to call interface with invalid hSteamUser`
> 
> That is the error. That is what makes the game crash before even getting to the port check
> 
> Also a **running** server works completely fine. I said you need to **restart** your server multiple times And you will see that at some starts it will not boot but throw this error
> 
> thats what i meant with flaky
> 
> Edit: It is not really a crash but more of an Exception. Seems to be some kind of try catch that will just output "Failed to get public ip from Steam" at any error

this pic and erropr tells me, that you have a problem with your network :D Not more

**Comment by rathmerdominik** [comment]
> > > > 
> > > 
> > > 
> > > i read it; my server runs now for 14h; since release and 2 hosters with 200+, customers have no problems; The servercheck tell you whats wrong; egg works fine
> > 
> > 
> > image
> > As you can read in the error message that i have sent you there is no information at all. It is a simple crash with an exception message that clearly states `Assertion Failed: Attempt to call interface with invalid hSteamUser`
> > That is the error. That is what makes the game crash before even getting to the port check
> > Also a **running** server works completely fine. I said you need to **restart** your server multiple times And you will see that at some starts it will not boot but throw this error
> > thats what i meant with flaky
> > Edit: It is not really a crash but more of an Exception. Seems to be some kind of try catch that will just output "Failed to get public ip from Steam" at any error
> 
> this pic and erropr tells me, that you have a problem with your network :D Not more

How is "Assertion failed: Attempt to call interface with invalid hSteamUser" a net error?
Can you explain that please?

And can you also explain why at some runs it **works**
I even showed you my server listed in the dedicated servers list.

This needs investigation not a discussion.

**Comment by gOOvER** [comment]
> > > > > 
> > > > 
> > > > 
> > > > i read it; my server runs now for 14h; since release and 2 hosters with 200+, customers have no problems; The servercheck tell you whats wrong; egg works fine
> > > 
> > > 
> > > image
> > > As you can read in the error message that i have sent you there is no information at all. It is a simple crash with an exception message that clearly states `Assertion Failed: Attempt to call interface with invalid hSteamUser`
> > > That is the error. That is what makes the game crash before even getting to the port check
> > > Also a **running** server works completely fine. I said you need to **restart** your server multiple times And you will see that at some starts it will not boot but throw this error
> > > thats what i meant with flaky
> > > Edit: It is not really a crash but more of an Exception. Seems to be some kind of try catch that will just output "Failed to get public ip from Steam" at any error
> > 
> > 
> > this pic and erropr tells me, that you have a problem with your network :D Not more
> 
> How is "Assertion failed: Attempt to call interface with invalid hSteamUser" a net error? Can you explain that please?
> 
> And can you also explain why at some runs it **works** I even showed you my server listed in the dedicated servers list.
> 
> This needs investigation not a discussion.

come on; you are able to read? If not, your problem is your IP.

LAST 2 Lines; Start learnig reading. I'Äm out here; its a problem on your site, not an egg issue

**Comment by gOOvER** [comment]
@QuintenQVD0 ignore the post above; he is not able to read and fix his firewall / network

**Comment by rathmerdominik** [comment]
How can an **IP** be an issue?

that is an obvious try catch block that catches any errors and prints this error message at **any** exception

can someone reasonable that is able to read a stacktrace, please have a look at this?

If this gets merged without any proper testing this could turn out rather bad.
It is with 100% certainty not a network issue.

I have also tested it now on three different machines (My PC, My Server, Friends PC) in two different networks and got the same report.

The error clearly is not about the Egg, but has either something to do with the interaction of wine with the Dedicated Server or something else that is unknown to me.

**Comment by gOOvER** [comment]
> How can an **IP** be an issue?
> 
> that is an obvious try catch block that catches any errors and prints this error message at **any** exception
> 
> can someone reasonable that is able to read a stacktrace, please have a look at this?
> 
> If this gets merged without any proper testing this could turn out rather bad. It is with 100% certainty not a network issue.
> 
> I have also tested it now on three different machines (My PC, My Server, Friends PC) in two different networks and got the same report.
> 
> The error clearly is not about the Egg, but has either something to do with the interaction of wine with the Dedicated Server or something else that is unknown to me.

read your error and ask the SOFT devs. And now pls stop.

**Comment by gOOvER** [comment]
sry 4 german now, but maybe he understand it thenm:

SOTF hat ein super system eingeführt, welches den Serfver testet; jeder Fehler der angezeigt wird, hat mit DIR zu tun ;) Also denk langsam mal drüber nach, anstatt mir irgendwelchen nonsense zu erzählen. Wenn man kein Ahnung hat, bitte nicht anfangen zu diskuteren. Mich interessieren keine Home server ;) Fix deine scheiss

**Comment by HerrSammyDE** [comment]
We have tested the Egg and it works! 🤝

**Comment by gOOvER** [comment]
@quinten. i can fix the config url tomorrow; so let this PR open until tomorrow

**Comment by QuintenQVD0** [comment]
@gOOvER is it ready to be merged?

**Comment by gOOvER** [comment]
> @gOOvER is it ready to be merged?

i updated the urls. Now its ready to merge

**Review by QuintenQVD0 — APPROVED** [review]
LGTM! Thank you

**Comment by user-420** [comment]
Unsure if this is the correct place to post this, I couldn't manage to find the guide on asking for support.

Having an issue where the egg isn't port-forwarding the ports, all other eggs I've used have no issue in doing so, see attached screenshot.
image


Have verified that the ports are indeed not open with check-host, any suggestions?

**Comment by gOOvER** [comment]
> 

its exact what servercheck said; check your ports. egg works fine. And use Discord for support

**Pull request merged**